### PR TITLE
special speed zone sign layout for many countries

### LINKED ISSUE #2954: Maxspeed zone sign has no "zone" label in Poland, mismatches with actual signs in other places
I think that special graphic in max speed quests in Poland would be helpful and be less confusing for mappers.

Image from https://doba.pl/dsw/artykul/komentarze/26044/64/
image

### PR DESCRIPTION
**Pull request by matkoniecz**
changelog:

Speed limit quest - labels on speed limit zone sign are now country-specific (by @matkoniecz). Note that speed limit quest is disabled by default as it requires more involved survey, you can enable it in settings. This specific quest is near the bottom of the quest list.

------------------------------------------------------------------------------------------------------------------------

fixes #2954

It has quite irritating layout duplication, but maybe it is safe to assume that it will be stable?

And alternative would be having layouts "zone text above sign", "zone text below sign" modified in various countries and it may not be sufficient anyway.

Progress (all are handled, ones with checkbox are also tested):

- [x] Japan is not offering zone speed sign in selection
- [x] Italy (label on top "zona") https://en.wikipedia.org/wiki/File:Italian_traffic_signs_-_zona_30_kmh.svg
- [ ] North Macedonia (ЗОНА label on bottom)
- [x] make "no label" default if not done already, remove Poland override
- [x] Poland (no label)
- [x] Sweden (no label, yellow background - so just testing required) https://commons.wikimedia.org/wiki/File:1_2_77_3_(Swedish_road_sign).svg
- [x] Germany "ZONE" at bottom https://commons.wikimedia.org/wiki/File:274-1-51.jpg
- [ ] UK  "ZONE" at bottom https://commons.wikimedia.org/wiki/File:UK_traffic_sign_674.svg
- [ ] Slovakia ZÓNA at top https://commons.wikimedia.org/wiki/File:275.1-20_Z%C3%B3na_najvy%C5%A1%C5%A1ej_dovolenej_r%C3%BDchlosti.jpg
- [x] Bosnia and Herzegovina "ЗОНА /newline/ ZONA" at bottom https://commons.wikimedia.org/wiki/File:Bosnia_and_Herzegovina_road_sign_III-26.svg
- [ ] Czechia ZÓNA at top https://commons.wikimedia.org/wiki/File:CZ-IZ05a_Z%C3%B3na_s_dopravn%C3%ADm_omezen%C3%ADm_(%C4%8Dtvercov%C3%A1)_(2016).jpg
- [x] Finland - no label, just test that it is yellow https://commons.wikimedia.org/wiki/File:Finland_road_sign_363-30_(1987%E2%80%931995).svg
- [x] Iceland - no label, test yellow https://commons.wikimedia.org/wiki/File:Iceland_road_sign_B28.30.svg
- [ ] Macedonia ЗОНА at bottom - https://commons.wikimedia.org/wiki/File:Macedonia_road_sign_315.1.svg
- [ ] Greece ΠΕΡΙΟΧΗ at top - https://commons.wikimedia.org/wiki/File:Road-sign-p60.svg
- [ ] Serbia ЗОНА at bottom https://commons.wikimedia.org/wiki/File:Serbian_road_sign_III-78.svg
- [ ] China :( https://en.wikipedia.org/wiki/File:China_road_sign_%E7%A6%81_43.svg
- [ ] Mauritius ZONE at bottom https://commons.wikimedia.org/wiki/File:Mauritius_Road_Signs_-_Prohibitory_Sign_-_Speed_limit_zone.svg
- [x] Russia ЗОНА at top https://commons.wikimedia.org/wiki/File:5.31_(Road_sign).gif
- [ ] Australia AREA at bottom https://commons.wikimedia.org/wiki/File:Australia_road_sign_R4-10_(20).svg
- [ ] Denmark Zone at bottom https://commons.wikimedia.org/wiki/File:Denmark_road_sign_E68.4.svg
- [ ] Macau ZONA on top https://commons.wikimedia.org/wiki/File:Macau_road_sign_22e.svg
- [ ] Netherlands ZONE on top https://commons.wikimedia.org/wiki/File:Netherlands_traffic_sign_A1-60-ZB.svg
- [ ] Ukraine ЗОНА on top https://commons.wikimedia.org/wiki/File:Ua_3.31_prohibitory-maximum_speed_zone.svg
- [ ] Slovenia CONA on top https://commons.wikimedia.org/wiki/File:Slovenia_road_sign_III-29_(40).svg
- [x] Armenia ԹՈՏՒ https://commons.wikimedia.org/wiki/File:5.31_Armenia_road_sign.svg
- [ ] Switzerland ZONE on top (second version ignored) https://commons.wikimedia.org/wiki/File:CH-Vorschriftssignal-Zonensignal-1.svg https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit
- [ ] Liechtenstein - the same as Switzerland ( https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit )
- [ ] Denmark - ????????????????????????? https://commons.wikimedia.org/wiki/File:Denmark_road_sign_E53.svg very suspicious! https://en.wikipedia.org/wiki/Road_signs_in_Denmark - argh.
- [x] France ZONE on top https://commons.wikimedia.org/wiki/File:France_road_sign_B30_(30).svg
- [ ] Hungary ZÓNA on bottom https://commons.wikimedia.org/wiki/File:Hungary_road_sign_E-028.svg
- [ ] Moldova ZONA on top https://commons.wikimedia.org/wiki/File:Moldova_road_sign_5.56.1.svg
- [ ] Romania ZONA on bottom https://commons.wikimedia.org/wiki/File:RomanianTrafficSign_C44-SpeedLimitZone-030_2011.svg
- [ ] Austria Zone on bottom https://commons.wikimedia.org/wiki/File:Vorschriftszeichen_11a_(2010).svg
- [ ] Estonia ALA on top https://en.wikipedia.org/wiki/Road_signs_in_Estonia#/media/File:EE_traffic_sign-382.png
- [x] Norway Sone on top https://en.wikipedia.org/wiki/Road_signs_in_Norway#/media/File:Norwegian-road-sign-366.0.svg
- [ ] Luxembourg ZONE on top, yellow background. Lets ignore yellow. https://commons.wikimedia.org/wiki/File:Luxembourg_road_sign_H,1-1.svg
- [ ] Belgium ZONE on top https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit
- [x] Ireland only slow zone (that has a yellow background) https://github.com/streetcomplete/StreetComplete/pull/3359#issuecomment-946369389
- [ ] Latvia ZONA at bottom https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit
- [ ] Portugal zona at top https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit
- [ ] Spain ZONA on top https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit
- [x] Croatia ZONA at bottom https://hr.wikipedia.org/wiki/Prometni_znakovi#/media/Datoteka:Croatia_road_sign_C21.svg
- [x] Japan 区域 ここあら on bottom https://github.com/streetcomplete/StreetComplete/pull/3359#issuecomment-946369389
- [x] Sweden - has no zone signs https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit
- [ ] Bulgaria - has no zone signs https://github.com/streetcomplete/StreetComplete/pull/3359#issuecomment-946166429

### CODE REVIEW

**Comment by matkoniecz** [comment]
> nordic countries (yellow background) should already be taken care of

I have bad news here! It is not working anymore. I spotted it after my mcc localization of untranslatable zone string failed to work.

I will try to locate what broke this, I remember that it used to work.

If my memory and understanding of code is right - traffic signs in Sweden should have yellow background.

1429ab262f61e4d5e35d2145490ceb630d3f61d5 is broken, v32.0 is broken, v31.0 compiled with https://github.com/google/dagger/issues/1339#issuecomment-462125185 bugfix applied[0] and in Sweden traffic light are still white, v20.0 has not yet this implemented, v25.0 has unknown for now situation (I will continue bisection).

[0]
to fix

> error: package javax.annotation.processing does not exist

replace
```
    val daggerVersion = "2.14.1"
```

with

```
    val daggerVersion = "2.14.1"
    compileOnly("com.github.pengrad:jdk9-deps:1.0")
```


Earlier it seemed that databinding would be suitable for parametrized layouts but...

> android.databinding.tool.processing.ScopedException: [databinding] {"msg":"Found \u003clayout\u003e but data binding is not enabled.\n\nAdd buildFeatures.dataBinding \u003d true to your build.gradle to enable it."

**Comment by westnordost** [comment]
so in maxspeed quest it is also not working?

**Comment by matkoniecz** [comment]
Or maybe get back to custom layouts without mcc-dependented resources?

I admit that when I looked at commit that introduced `@color/traffic_yellow` and replaced per-area layouts I was unsure is it a significant improvements given that it relies on bunch of magic to work. I assumed that I probably miss something and that my aversion exists because I avoid learning about new solution.

But given that magic is actually not working and requires even more magic to work...

I would consider reverting https://github.com/streetcomplete/StreetComplete/commit/ec2529aa25ea5f9dc4731a161399e6e398592e6f as one more solution (I am not saying that it is preferable, but I would at least consider it as an option).

**Comment by westnordost** [comment]
Just look at all the `drawable-mcc***` resources, that would be a lot more work than to set the drawable manually in code after inflating, wouldn't you agree?

I mean, that the layouts are per mcc is also magic we rely on to work. If we were to abandon that, we'd need to get rif of **all** resources referring to MCCs but instead put this into code. A possibility of course, but again, this is a lot of work. And not sure what we gain with this.

**Comment by matkoniecz** [comment]
If you see anything wrong in https://github.com/streetcomplete/StreetComplete/pull/3359#issuecomment-946117709 list - please comment.

If you know how zone speed sign look for one of missing countries (or you know that country is without it) - also please comment.

**Comment by mnalis** [comment]
> If you see anything wrong in #3359 (comment) list - please comment.

I don't see Croatia in that list; we have text `ZONA` below sign:
https://hr.wikipedia.org/wiki/Prometni_znakovi#/media/Datoteka:Croatia_road_sign_C21.svg

**Comment by matkoniecz** [comment]
@mnalis @Kovoschiz Thanks!

> https://commons.wikimedia.org/wiki/File:Ireland_road_sign_F_403.svg is only used together with a wrong living_street-style designation (no pedestrian priority), pictogram at top, "crios mall slow zone" at bottom. But apparently there are some non-compliant examples with the order swapped, one at https://learnirishfromroadsigns.art.blog/2020/03/20/crios-mall-slow-zone/ .

As slow zone is handled separately, than means that Ireland de facto has no zone speed limits.

Also, it will be worth checking is there already support for yellow background on living zone traffic sign.

----------------------------------------------

Note: I will likely keep this list for some time when I have time but are for some reason unable to handle something more complex as it is mostly drudgery with translating country names to https://en.wikipedia.org/wiki/Mobile_country_code and creating bunch of layouts and texts, like in https://github.com/streetcomplete/StreetComplete/blob/caad4bfbc1378ee813f626a467a12d189bf79cc5/app/src/main/res/values-mcc294/strings.xml https://github.com/streetcomplete/StreetComplete/blob/caad4bfbc1378ee813f626a467a12d189bf79cc5/app/src/main/res/layout-mcc294/quest_maxspeed_zone_sign.xml

**Comment by riiga** [comment]
>Sweden (no label, yellow background - so just testing required) https://commons.wikimedia.org/wiki/File:1_2_77_3_(Swedish_road_sign).svg
>Sweden - has no zone signs https://en.wikipedia.org/wiki/Comparison_of_European_road_signs#Speed_limit

Sweden has a sign for speed zones, but doesn't use it. I've never seen any other zones apart from [no parking](https://transportstyrelsen.se/globalassets/global/vag/vagmarken/vagmarken-nedladdning/e20-1.png)/[no stopping](https://transportstyrelsen.se/globalassets/global/vag/vagmarken/vagmarken-nedladdning/e20-15.png) zones and the occasional [parking zone](https://transportstyrelsen.se/globalassets/global/vag/vagmarken/vagmarken-nedladdning/e20-16.png). We don't have any signs for default speed limits either, all speed limits are signed with a regular speed limit sign. The exception are [recommended lower speed limits](https://transportstyrelsen.se/globalassets/global/vag/vagmarken/vagmarken-nedladdning/e11-1.png) which use [an end sign](https://transportstyrelsen.se/globalassets/global/vag/vagmarken/vagmarken-nedladdning/e12-1.png).

>Denmark Zone at bottom https://commons.wikimedia.org/wiki/File:Denmark_road_sign_E68.4.svg
>Denmark - ????????????????????????? https://commons.wikimedia.org/wiki/File:Denmark_road_sign_E53.svg very suspicious! https://en.wikipedia.org/wiki/Road_signs_in_Denmark - argh.

Denmark has two different signs for speed zones. The blue one is for areas where there is traffic calming and the the speed is generally 30 or 20 km/h. The regular one is for what you'd expect.

**Comment by FloEdelmann** [comment]
Or maybe a file `maxSpeedZone.yml` in [`res/country_metadata/`](https://github.com/streetcomplete/StreetComplete/tree/master/res/country_metadata)?

```yaml
default: false # no speed limit zones known by default
DE: "ZONE"
HT: "ZÓNA"
…
```

**Comment by matkoniecz** [comment]
What about:

`speedZoneLabel.yml`

```
PL: none
DE: {text: ZONE, position: below}
```

**Comment by westnordost** [comment]
> speedZoneLabel.yml

> default: none
> DE: {text: ZONE, position: below}

No, the files in country-metadata should be flat. But they can be named similarily.

`slowZoneLabelPosition.yml` + `slowZoneLabelText.yml`



>> Bulgaria doesn't have such signs since there is no definition in the traffic law.

> It would be nice to somehow handle this

https://github.com/streetcomplete/StreetComplete/blob/master/res/country_metadata/isSlowZoneKnown.yml

**Comment by matkoniecz** [comment]
@westnordost 

> No, the files in country-metadata should be flat. But they can be named similarily.

what about

```
DE:
  - ZONE
  - bottom_of_sign
SI:
  - CONA
  - top_of_sign
```
?

A bit cryptic, but still far nicer to edit than two separate files.

**Comment by westnordost** [comment]
No, for consistency the files should be flat.

**Comment by westnordost** [comment]
What exactly is missing here? Is it ready for review&merge? In the list it shows up as 16/46 tasks finished.

File: app/src/main/java/de/westnordost/streetcomplete/quests/max_speed/AddMaxSpeedForm.kt [hunk 1]
```diff
@@ -121,6 +127,32 @@ class AddMaxSpeedForm : AbstractQuestFormAnswerFragment<MaxSpeedAnswer>() {
         else -> null
     }
 
+    private fun enableAppropriateLabelsForSlowZone(layoutWithSign: FrameLayout) {
+        val position = countryInfo.slowZoneLabelPosition
+        val text = countryInfo.slowZoneLabelText
+
+        // no label
+        if (position == null && text == null) {
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/max_speed/AddMaxSpeedForm.kt:135]: Please remove these strict checks from the code. This is just the UI, I'd rather have the zone sign not display 100% correctly than crash the app if data is not as expected. E.g. if position is not as expected, just use "top". If text is null, just leave the label visibility on GONE.

```kotlin
val label: MaterialTextView = layoutWithSign.findViewById(
    if (position == "bottom") R.id.slowZoneLabelBottom else R.id.slowZoneLabelTop
)
label.visibility.isGone = countryInfo.slowZoneLabelText == null
label.text = countryInfo.slowZoneLabelTex
```

**Review by westnordost — CHANGES_REQUESTED** [review]

**Comment by westnordost** [comment]
> Though if you think that it would be useful I can also test in remaining 30 locations.

Nah that's fine

File: app/src/main/java/de/westnordost/streetcomplete/quests/max_speed/AddMaxSpeedForm.kt (hunk 1 above)
Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/quests/max_speed/AddMaxSpeedForm.kt:135]: Removed.

**Review by westnordost — APPROVED** [review]

**Comment by westnordost** [comment]
Since no new strings were added and you tested this well, this might as well go into a minor release (in case there'll be one), so we can merge it now

**Comment by matkoniecz** [comment]
@westnordost if you want you can use following as a changelog entry

> Speed limit quest - labels on speed limit zone sign are now country-specific (by @matkoniecz). Note that speed limit quest is disabled by default as it requires more involved survey, you can enable it in settings. This specific quest is near the bottom of the quest list.

(I can also add it directly if you think that it is fine)

**Comment by smichel17** [comment]
Cleaned up:

> Speed limit quest - labels on speed limit zone sign are now country-specific (by @matkoniecz). Note: the speed limit quest is disabled by default as it requires more involved survey. You can enable it in settings, near the bottom of the quest list.

**Comment by westnordost** [comment]
Regarding updates on the changelog:
- I'd rather completely do it myself
- **OR** if someone else updates the changelog, that person should add **all the things** (relevant for users) that have been added since the last time the changelog was updated.

To keep the overview till which commit the changelog was maintained, you know.

In general, the changelog entries should be **very short** / simplified. For more information and notes, there is always the link to the actual ticket. So for this PR, I'd rather add someting like "display speed limit zone signs as they appear in your country in the form" or similar.

**Comment by westnordost** [comment]
> Regarding updates on the changelog:

Any anyone with write acces can do this, without a PR, if he follows these rules. I wanted to put this into the "working together efficiently" post in the discussion forum but then again I thought this is not going to be a big effort for me as I need to go through the commit history anyway before each release again to see if hte changelog is up-to-date.

**Comment by Dimitar5555** [comment]
> Bulgaria doesn't have such signs since there is no definition in the traffic law.

A few months ago this sign was introduced. It has "ЗОНА" on the top side.

image

**Comment by matkoniecz** [comment]
done in https://github.com/streetcomplete/countrymetadata/pull/8

in general, for such thing it is better to open new issue - closed PRs and issues can be no longer watched, and if not processed immediately it may be forgotten

**Pull request merged**